File: Lab4/TestLearner.java
package Lab4;

public class TestLearner {
    public static void main(String[] args) {
        System.out.println("INICIO DEL TEST DE APRENDIZAJE Y ABSTRACCIÓN \n");
        
        RawDataset dataset = new RawDataset(2);
        //Queremos entrenar a nuestro modelo para que sume los componentes del vector, y que ese sea el Record del Vector en cuestión
        dataset.addRecord(new Record(new Vector(new double[]{1, 1}), 2));   
        dataset.addRecord(new Record(new Vector(new double[]{2, 3}), 5));  
        dataset.addRecord(new Record(new Vector(new double[]{4, 1}), 5));   
        dataset.addRecord(new Record(new Vector(new double[]{0, 0}), 0));   
        dataset.addRecord(new Record(new Vector(new double[]{-1, 1}), 0));  
        dataset.addRecord(new Record(new Vector(new double[]{10, 5}), 15)); 

        System.out.println("Datos cargados: " + dataset.getData().size() + " registros.");

        System.out.println("\n Estadísticas del RawDataset ");
        System.out.println("Media Input: " + dataset.meanInput());
        System.out.println("Media Output: " + dataset.meanOutput());

        StandardizedDataset standardizedDataset = dataset.standardize();

        System.out.println("Dataset estandarizado creado correctamente.");

        System.out.println("\n\n PRUEBA 1: Abstracción con Gradient Descent ");
        
        Algorithm algoritmo1 = new GradientDescent(0.01, 0.001);
        
        System.out.println("Variable 'algoritmo1' declarada como tipo: Algorithm");
        
        SupervisedLearner learnerGD = new SupervisedLearner(algoritmo1, standardizedDataset);
        
        learnerGD.solve();
        System.out.println("Modelo entrenado (Pesos): " + learnerGD.toString());

        Vector vector_predict = new Vector(new double[]{10, 10});
        double prediccionGD = learnerGD.predict(vector_predict);
        
        System.out.println("Pregunta: 10 + 10");
        System.out.println("Predicción GD: " + prediccionGD);
        System.out.println("Error: " + Math.abs(20 - prediccionGD));

        System.out.println("\n\n PRUEBA 2: Abstracción con Stochastic Gradient Descent ");

        Algorithm algoritmo2 = new StochasticGradientDescent(0.01, 2, 1000);
        
        System.out.println("Variable 'algoritmo2' declarada como tipo: Algorithm");

        SupervisedLearner learnerSGD = new SupervisedLearner(algoritmo2, standardizedDataset);
        
        learnerSGD.solve();
        System.out.println("Modelo SGD entrenado (Pesos): " + learnerSGD.toString());

        // PREDICCIÓN
        double prediccionSGD = learnerSGD.predict(vector_predict);
        
        System.out.println("Pregunta: 10 + 10");
        System.out.println("Predicción SGD: " + prediccionSGD);
        System.out.println("Error: " + Math.abs(20 - prediccionSGD));

        System.out.println("\n=== FIN DEL TEST ===");
    } 
    
}

File: Lab5/students/TestLearner.cpp
#include <iostream>
#include "Dataset.h"
#include "Model.h"
#include "GradientDescent.h"          
#include "StochasticGradientDescent.h" 
#include "SupervisedLearner.h"        

int main() {


    Vector v1( std::vector<double> { 0.5, 1.5, 2.5 } );
    Vector v2( std::vector<double> { 8.0, 6.0, 4.0 } );
    Vector v3( std::vector<double> { 3.6, 4.8, 7.2 } );
    
    std::cout << v1 << "\n" << v2 << "\n" << v3 << "\n";
    
    Vector v4 = v1.add( v3 );
    Vector v5 = v2.subtract( v1 );
    Vector v6 = v1.multiply( 4 );
    Vector v7 = v3.divide( 1.2 );
    Vector v8 = v2.multiply( v2 );
    
    std::cout << v4 << "\n" << v5 << "\n" << v6 << "\n" << v7 << "\n" << v8 << "\n";
    
    double d1 = v1.dotProduct( v2 );
    double d2 = v2.norm();
    
    std::cout << d1 << " " << d2 << "\n";

    //y = 2x
    Dataset ds(1);
    ds.addRecord(Record(Vector(std::vector<double>{1.0}), 2.0));
    ds.addRecord(Record(Vector(std::vector<double>{2.0}), 4.0));
    ds.addRecord(Record(Vector(std::vector<double>{3.0}), 6.0));
    ds.addRecord(Record(Vector(std::vector<double>{4.0}), 8.0));


    std::cout << "\n--> Probando Gradient Descent:\n";
    // lr=0.01, sc=0.0001
    GradientDescent* gd = new GradientDescent(0.01, 0.0001);
    SupervisedLearner learnerGD(gd, ds);
    
    learnerGD.solve(); 
    
    // Probamos predicción para 5 (debería dar 10)
    Vector query(std::vector<double>{5.0});
    std::cout << "Prediccion para 5.0: " << learnerGD.predict(query) << "\n";
    
    delete gd;


    std::cout << "\n--> Probando Stochastic Gradient Descent:\n";
    // lr=0.01, batchSize=2, iterations=1500
    StochasticGradientDescent* sgd = new StochasticGradientDescent(0.01, 2, 1500);
    SupervisedLearner learnerSGD(sgd, ds);
    
    learnerSGD.solve(); 
    
    std::cout << "Prediccion para 5.0: " << learnerSGD.predict(query) << "\n";
    delete sgd;

    return 0;
}

